# Contents
01- Import Library
02- Import Data
03- Task 4.8 Grouping and Aggregating Data
    3.1 Aggregating variables
    3.2 Grouping and aggregating loyalty customers
    3.3 Aggregating average prices
    3.4 Grouping a spending flag
    3.5 Aggregating median order frequency
    3.6 Grouping a frequency customer order flag
04- Saving Data frame

# 01 Import Library

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

# 02 import data

In [2]:
#folder shortcut
path = r'/Users/rachelmoreland/06_2023_Instacart Basket Analysis'

In [4]:
# import orders_prods_merged
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_47.pkl'))

In [5]:
df_ords_prods_merge.shape

(32404859, 20)

In [6]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     0.0                 10          0   both   
4                     9.0                 11          1   both   

         price_label     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most Orders   
1  Mid-range product  Regularly busy  Regularly busy        Average Orders   
2  Mid-range product     Busiest day    Busiest days        Average Orders   
3  Mid-range product  Regularly busy      Least busy           Most Orders   
4  Mid-range product      Least busy      Least busy        Average Orders   

   max_order      loyalty_flag  
0         32  Regular customer  
1         32  Regular customer  
2          5      New customer  
3          3      New customer  
4          3      New customer

# 03 Task

## 3.1 aggregating variables

In [7]:
#2 finding the agg mean of order number grouped by dept id

df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## 3 comparing these results with the subset I first see that dept id 21 is less that in the subset ( subset was 25.5 and whole data set is 22.9) more than 2 dollars. there is also a bit of a difference with dept id 10.  Aside from that the trends remain pretty similar.

## 3.2 Grouping loyalty customers and checking aggregations

In [10]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     0.0                 10          0   both   
4                     9.0                 11          1   both   

         price_label     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most Orders   
1  Mid-range product  Regularly busy  Regularly busy        Average Orders   
2  Mid-range product     Busiest day    Busiest days        Average Orders   
3  Mid-range product  Regularly busy      Least busy           Most Orders   
4  Mid-range product      Least busy      Least busy        Average Orders   

   max_order  
0         32  
1         32  
2          5  
3          3  
4          3

In [11]:
#4 creating loyalty flag column

df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
#checking freq counts
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [15]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     0.0                 10          0   both   
4                     9.0                 11          1   both   

         price_label     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most Orders   
1  Mid-range product  Regularly busy  Regularly busy        Average Orders   
2  Mid-range product     Busiest day    Busiest days        Average Orders   
3  Mid-range product  Regularly busy      Least busy           Most Orders   
4  Mid-range product      Least busy      Least busy        Average Orders   

   max_order      loyalty_flag  
0         32  Regular customer  
1         32  Regular customer  
2          5      New customer  
3          3      New customer  
4          3      New customer

In [16]:
#5 seeing if there is a spending difference based on type of customer
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

## it looks like loyal customers find a way to spend less on the items they are purchasing

## 3.3 Aggregating average prices 

In [17]:
#6 grouped by 'user_id' and 'prices' transformed into mean of 'prices' making new column avg prices

df_ords_prods_merge['avg_prices'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [18]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  add_to_cart_order  reordered  _merge        price_label  \
0  ...                  5          0    both  Mid-range product   
1  ...                  1          1    both  Mid-range product   
2  ...                 20          0    both  Mid-range product   
3  ...                 10          0    both  Mid-range product   
4  ...                 11          1    both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy           Most Orders        32   
1  Regularly busy  Regularly busy        Average Orders        32   
2     Busiest day    Busiest days        Average Orders         5   
3  Regularly busy      Least busy           Most Orders         3   
4      Least busy      Least busy        Average Orders         3   

       loyalty_flag avg_prices  
0  Regular customer   6.935811  
1  Regular customer   6.935811  
2      New customer   7.930208  
3      New customer   4.972414  
4      New customer   4.972414  

[5 rows x 21 columns]

## 3.4 grouping a spending flag

In [19]:
#6 creating spending flag column
df_ords_prods_merge.loc[df_ords_prods_merge['avg_prices'] < 10, 'spending_flag'] = 'Low Spender'

In [20]:
df_ords_prods_merge.loc[df_ords_prods_merge['avg_prices'] >= 10, 'spending_flag'] = 'High Spender'

In [21]:
#checking freq counts
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low Spender     31770614
High Spender      634245
Name: spending_flag, dtype: int64

In [22]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  reordered  _merge        price_label     busiest_day    busiest_days  \
0  ...          0    both  Mid-range product  Regularly busy  Regularly busy   
1  ...          1    both  Mid-range product  Regularly busy  Regularly busy   
2  ...          0    both  Mid-range product     Busiest day    Busiest days   
3  ...          0    both  Mid-range product  Regularly busy      Least busy   
4  ...          1    both  Mid-range product      Least busy      Least busy   

  busiest_period_of_day max_order      loyalty_flag  avg_prices spending_flag  
0           Most Orders        32  Regular customer    6.935811   Low Spender  
1        Average Orders        32  Regular customer    6.935811   Low Spender  
2        Average Orders         5      New customer    7.930208   Low Spender  
3           Most Orders         3      New customer    4.972414   Low Spender  
4        Average Orders         3      New customer    4.972414   Low Spender  

[5 rows x 22 columns]

## 3.5 Aggregating median order frequency

In [23]:
#7 grouped by 'user_id' and 'days since prior' transformed into median order freq making new column 

df_ords_prods_merge['median_order_freq'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [24]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  _merge        price_label     busiest_day    busiest_days  \
0  ...    both  Mid-range product  Regularly busy  Regularly busy   
1  ...    both  Mid-range product  Regularly busy  Regularly busy   
2  ...    both  Mid-range product     Busiest day    Busiest days   
3  ...    both  Mid-range product  Regularly busy      Least busy   
4  ...    both  Mid-range product      Least busy      Least busy   

  busiest_period_of_day max_order      loyalty_flag avg_prices  spending_flag  \
0           Most Orders        32  Regular customer   6.935811    Low Spender   
1        Average Orders        32  Regular customer   6.935811    Low Spender   
2        Average Orders         5      New customer   7.930208    Low Spender   
3           Most Orders         3      New customer   4.972414    Low Spender   
4        Average Orders         3      New customer   4.972414    Low Spender   

  median_order_freq  
0               8.0  
1               8.0  
2               6.0  
3               9.0  
4               9.0  

[5 rows x 23 columns]

In [25]:
df_ords_prods_merge.shape

(32404859, 23)

## 3.6 grouing a frequency customer order flag

In [26]:
#7 creating a freq customer flag
df_ords_prods_merge.loc[df_ords_prods_merge['median_order_freq'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [27]:
df_ords_prods_merge.loc[(df_ords_prods_merge['median_order_freq'] <= 20) & (df_ords_prods_merge['median_order_freq'] > 10), 'frequency_flag'] = 'Regular customer'

In [28]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_order_freq'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [29]:
#checking freq counts
df_ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: frequency_flag, dtype: int64

In [30]:
df_ords_prods_merge[['user_id', 'loyalty_flag', 'spending_flag', 'frequency_flag','order_number']].head(60).sort_values('user_id')

user_id      loyalty_flag spending_flag         frequency_flag  \
0       138  Regular customer   Low Spender      Frequent customer   
1       138  Regular customer   Low Spender      Frequent customer   
2       709      New customer   Low Spender      Frequent customer   
3       764      New customer   Low Spender      Frequent customer   
4       764      New customer   Low Spender      Frequent customer   
5       777  Regular customer   Low Spender       Regular customer   
6       825      New customer   Low Spender       Regular customer   
7       910  Regular customer   Low Spender      Frequent customer   
8      1052  Regular customer   Low Spender      Frequent customer   
9      1052  Regular customer   Low Spender      Frequent customer   
10     1379      New customer   Low Spender       Regular customer   
13     1480    Loyal customer   Low Spender      Frequent customer   
12     1480    Loyal customer   Low Spender      Frequent customer   
11     1480    Loyal customer   Low Spender      Frequent customer   
14     1494  Regular customer   Low Spender      Frequent customer   
15     1494  Regular customer   Low Spender      Frequent customer   
16     1494  Regular customer   Low Spender      Frequent customer   
33     1540  Regular customer   Low Spender      Frequent customer   
32     1540  Regular customer   Low Spender      Frequent customer   
31     1540  Regular customer   Low Spender      Frequent customer   
30     1540  Regular customer   Low Spender      Frequent customer   
28     1540  Regular customer   Low Spender      Frequent customer   
27     1540  Regular customer   Low Spender      Frequent customer   
26     1540  Regular customer   Low Spender      Frequent customer   
25     1540  Regular customer   Low Spender      Frequent customer   
29     1540  Regular customer   Low Spender      Frequent customer   
23     1540  Regular customer   Low Spender      Frequent customer   
22     1540  Regular customer   Low Spender      Frequent customer   
21     1540  Regular customer   Low Spender      Frequent customer   
24     1540  Regular customer   Low Spender      Frequent customer   
20     1540  Regular customer   Low Spender      Frequent customer   
19     1540  Regular customer   Low Spender      Frequent customer   
18     1540  Regular customer   Low Spender      Frequent customer   
17     1540  Regular customer   Low Spender      Frequent customer   
34     1598      New customer   Low Spender      Frequent customer   
35     1598      New customer   Low Spender      Frequent customer   
36     1598      New customer   Low Spender      Frequent customer   
37     1647      New customer   Low Spender  Non-frequent customer   
38     1647      New customer   Low Spender  Non-frequent customer   
39     1647      New customer   Low Spender  Non-frequent customer   
40     1647      New customer   Low Spender  Non-frequent customer   
41     1647      New customer   Low Spender  Non-frequent customer   
42     2833      New customer   Low Spender      Frequent customer   
43     2850  Regular customer   Low Spender      Frequent customer   
44     2857  Regular customer   Low Spender      Frequent customer   
45     3010  Regular customer   Low Spender      Frequent customer   
46     3029      New customer   Low Spender  Non-frequent customer   
47     3395      New customer   Low Spender       Regular customer   
48     3904      New customer   Low Spender       Regular customer   
53     4006    Loyal customer   Low Spender      Frequent customer   
52     4006    Loyal customer   Low Spender      Frequent customer   
51     4006    Loyal customer   Low Spender      Frequent customer   
50     4006    Loyal customer   Low Spender      Frequent customer   
49     4006    Loyal customer   Low Spender      Frequent customer   
54     4122  Regular customer   Low Spender      Frequent customer   
55     4141      New customer   Low Spender  Non-frequent customer   
56     4343   

In [31]:
df_ords_prods_merge.shape

(32404859, 24)

# 04 saving data frame

In [32]:
# Export data to pkl

df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_48.pkl'))

In [ ]:
#changing the prices of items over 100 to nan
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] >100, 'prices'] = np.nan